src/Seabug/Contract/MarketPlaceListNft.purs

```diff
@@ -1,59 +1,64 @@
-import Data.Array (catMaybes)
-import Data.Identity (Identity(Identity))
+import Control.Monad.Reader (asks)
+import Control.Parallel (parTraverse)
+import Data.Array (catMaybes, mapMaybe)
 import Data.Map as Map
-import Plutus.ToPlutusType (toPlutusType)
+import Seabug.Contract.Common (NftResult)
+import Seabug.Contract.Util (minAdaOnlyUTxOValue)
 import Seabug.MarketPlace (marketplaceValidator)
-import Seabug.Metadata (FullSeabugMetadata, getFullSeabugMetadata)
+import Seabug.Metadata (getFullSeabugMetadataWithBackoff)
 import Seabug.Types (MarketplaceDatum(MarketplaceDatum))
 
-type ListNftResult =
-  { input :: TransactionInput
-  , output :: TransactionOutput
-  , metadata :: FullSeabugMetadata
-  }
-
 -- | Lists the utxos at the script address that contain a datum of type
 -- | `MarketplaceDatum` with unit value. It currently doesn't have any logic
 -- | on matching `CurrencySymbol` and `TokenName`.
 marketPlaceListNft
   :: forall (r :: Row Type)
-   . Contract (projectId :: String | r) (Array ListNftResult)
+   . Contract (projectId :: String | r) (Array NftResult)
 marketPlaceListNft = do
   marketplaceValidator' <- unwrap <$> liftContractE marketplaceValidator
   networkId <- getNetworkId
-  let
-    scriptAddr =
-      typedValidatorEnterpriseAddress networkId $ wrap marketplaceValidator'
+  projectId <- asks $ unwrap >>> _.projectId
+  scriptAddr <-
+    liftedM "marketPlaceListNft: Cannot convert validator hash to address"
+      $ pure
+      $ typedValidatorEnterpriseAddress networkId
+      $ wrap marketplaceValidator'
   scriptUtxos <- Map.toUnfoldable <<< unwrap <$>
-    liftedM "marketPlaceListNft: Cannot get script Utxos" (utxosAt scriptAddr)
-  withMetadata <- for scriptUtxos $
+    liftedM "marketPlaceListNft: Cannot get script Utxos"
+      (utxosAt scriptAddr)
+  datums <- getDatumsByHashes
+    $ mapMaybe (snd >>> unwrap >>> _.dataHash) scriptUtxos
+  withMetadata <- liftAff $ (flip parTraverse) scriptUtxos $
     \(input /\ output@(TransactionOutput out)) ->
       runMaybeT $ do
-        datumHash <- MaybeT $ pure $ out.dataHash
-        plutusData <- MaybeT $ getDatumByHash datumHash
         MarketplaceDatum { getMarketplaceDatum: curr /\ name } <-
-          MaybeT $ pure $ fromData plutusData
-        let (Identity amt) = toPlutusType out.amount
-        guard $ valueOf amt curr name == one
-        metadata <- MaybeT $ map hush $ getFullSeabugMetadata $ curr /\ name
+          MaybeT $ pure $ (fromData <<< unwrap)
+            =<< (_ `Map.lookup` datums)
+            =<< out.dataHash
+        guard $ valueOf out.amount curr name == one
+        metadata <- MaybeT $ map hush $
+          getFullSeabugMetadataWithBackoff (curr /\ name) projectId
+        -- TODO: this is a temporary solution to only show NFTs known
+        -- to work. This filter catches a couple new nfts whose price
+        -- I put too low. The old nfts are caught above because their
+        -- metadata won't be parsed.
```

Comment:
While this guard does make sense, we'll likely drop this logic in the future, as the filtering will be handled by the backend. For now though its fine.